File: object-oriented-programming/object_oriented_programming.py
from abc import ABC, abstractmethod


class Animal(ABC):
    """Abstract base class demonstrating abstraction."""
    
    def __init__(self, name: str):
        self.name = name
    
    @abstractmethod
    def speak(self) -> str:
        pass
    
    def describe(self) -> str:
        return f"I am {self.name}"


class Dog(Animal):
    """Concrete class inheriting from abstract base."""
    
    def speak(self) -> str:
        return f"{self.name} says: Woof!"


class Cat(Animal):
    """Another concrete class with method override."""
    
    def speak(self) -> str:
        return f"{self.name} says: Meow!"


def main() -> None:
    print("Language: Python")
    
    animals: list[Animal] = [Dog("Rex"), Cat("Whiskers")]
    
    for animal in animals:
        print(animal.speak())
        print(animal.describe())


if __name__ == "__main__":
    main()

File: object-oriented-programming/ObjectOrientedProgramming.cpp
#include <iostream>
#include <memory>
#include <string>
#include <vector>

// Abstract base class
class Animal {
public:
    virtual ~Animal() = default;

    virtual std::string speak() const = 0;

    virtual std::string describe() const {
        return "I am " + name;
    }

protected:
    std::string name;

    Animal(const std::string& n) : name(n) {}
};

// Concrete class: Dog
class Dog : public Animal {
public:
    Dog(const std::string& name) : Animal(name) {}

    std::string speak() const override {
        return name + " says: Woof!";
    }
};

// Concrete class: Cat
class Cat : public Animal {
public:
    Cat(const std::string& name) : Animal(name) {}

    std::string speak() const override {
        return name + " says: Meow!";
    }
};

int main() {
    std::cout << "Language: C++\n";

    std::vector<std::unique_ptr<Animal>> animals;
    animals.push_back(std::make_unique<Dog>("Rex"));
    animals.push_back(std::make_unique<Cat>("Whiskers"));

    for (const auto& animal : animals) {
        std::cout << animal->speak() << "\n";
        std::cout << animal->describe() << "\n";
    }

    return 0;
}
